### PR TITLE
Prevent training from ending during warmup

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

- honor `warmup_epochs` as an epoch count instead of forcing at least 100 iterations
- cap warmup before the final planned epoch so training always reaches the regular LR schedule
- centralize warmup iteration calculation for initial setup, OOM batch-size rebuilds, and timed-training epoch re-estimates
- bump version to 8.4.103

A 10-epoch run with 9 batches per epoch previously spent all 90 iterations inside the hard 100-iteration warmup. It now warms up for the configured 3 epochs (27 iterations) and trains normally for the remaining 7.

## Validation

No test additions — verified empirically on real CPU trainings (yolo11n, coco8, imgsz=32, batch=2, nb=2):

- epochs=10, warmup_epochs=3 → warmup active for exactly nw=6 iterations (accumulate 1→27, lr 0→8e-5), then the regular linear schedule through the final epoch. On main the same config runs ALL 20 iterations inside the 100-floor warmup: peak lr 6.6e-6 (~18x below nominal), accumulate never reaches its target, and the run ends severely under-trained.
- epochs=1 → nw=0, zero warmup, full nominal lr from the first batch, finite loss (`self.accumulate` is pre-initialized in `_build_train_pipeline`, so skipping warmup leaves nothing undefined).
- warmup_epochs=0 → nw=0, exactly equivalent to the old -1 sentinel (`nw` is consumed only at the `ni < nw` check).
- timed mode (`args.time`): nw now tracks the re-estimated epoch count, so a shortened run also shortens warmup instead of keeping a stale bound.
- No subclass overrides `_do_train`/`_setup_train` or computes `nw`; no doc documents the old 100-iteration floor.
- `ruff format` / `ruff check` clean.

Deleted: the undocumented 100-iteration warmup minimum, the `-1` no-warmup sentinel, the inclusive `ni <= nw` boundary, and the two duplicated inline `nw` computations (initial setup + OOM rebuild), consolidated into one `_get_warmup_iterations()` helper.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Improves training warmup scheduling to prevent warmup from consuming the entire training run and keep it consistent when training duration changes. 🔧

### 📊 Key Changes

- Added `_get_warmup_iterations()` to calculate warmup steps while reserving at least the final epoch for regular training.
- Removed the previous minimum of 100 warmup iterations, allowing short training runs to use an appropriate warmup duration.
- Corrected the warmup boundary condition so warmup ends precisely at the calculated iteration.
- Recalculates warmup iterations when:
  - The batch size is reduced and the training pipeline is rebuilt.
  - Time-based training updates the total number of epochs.
- Incremented the Ultralytics package version from `8.4.102` to `8.4.103`.

### 🎯 Purpose & Impact

- Prevents warmup from extending through the entire training process during very short runs. ⏱️
- Makes learning-rate and gradient-accumulation warmup behavior more predictable and consistent.
- Improves resilience when dynamic batch-size adjustment or time-limited training changes the training schedule.
- Helps models transition to regular training sooner, potentially improving convergence and training efficiency. 🚀

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

This is an automated message, and an engineer will assist with any questions or follow-up needed. 🛠️

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -395,7 +400,7 @@ def _do_train(self):
         self._setup_train()
 
         nb = len(self.train_loader)  # number of batches
-        nw = max(round(self.args.warmup_epochs * nb), 100) if self.args.warmup_epochs > 0 else -1  # warmup iterations
+        nw = self._get_warmup_iterations(nb)
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:403]: 💡 **MEDIUM**: This is calculated only once, but timed training recomputes `self.epochs` after every epoch (`self.epochs = ...` in the `args.time` branch). A short `time=` run can therefore finish in fewer epochs than the initially calculated warmup (typically 3), so it still never reaches the regular LR schedule. Recalculate `nw` whenever the timed branch updates the planned epoch count, and add a `time`-mode regression test.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The fixed-epoch warmup behavior and OOM rebuild path are correctly centralized, but timed training can still consume the entire run in warmup because its dynamically updated epoch count does not refresh `nw`.

💬 Posted 1 inline comment

File: ultralytics/engine/trainer.py (hunk 1 above)
Comment (glenn-jocher) [reply ultralytics/engine/trainer.py:403]: Fixed in 008527512. The timed-training branch now recalculates `nw` immediately after updating its epoch estimate, and `test_warmup_iterations_follow_timed_epoch_estimate` covers the estimate shrinking from 100 epochs to 2.

**Comment by glenn-jocher** [comment]
## Controlled before/after training

Compared `origin/main` (`f3c0d1816`) against PR head `008527512` with identical pretrained YOLO26n weights, data, `epochs=10`, seed 0, automatic optimizer, `imgsz=160`, and Apple M4 Pro MPS. COCO128 used `batch=64` (2 batches/epoch); DOTA8 used `batch=8` (1 batch/epoch).

| Dataset / task | Effective warmup | Peak LR (pg0) | Final cls loss | Final mAP50 | Other final metric |
|---|---:|---:|---:|---:|---:|
| COCO128 detect, before | 20/20 steps | 6.61e-6 | 4.240 | 0.269 box | 0.170 box mAP50-95 |
| COCO128 detect, after | 6/20 steps | 8.37e-5 | 2.947 | 0.311 box | 0.181 box mAP50-95 |
| COCO128-seg, before | 20/20 steps | 6.61e-6 | 4.247 | 0.272 box | 0.244 mask mAP50 |
| COCO128-seg, after | 6/20 steps | 8.37e-5 | 3.030 | 0.319 box | 0.270 mask mAP50 |
| DOTA8 OBB, before | 10/10 steps | 1.33e-5 | 6.511 | 0.332 box | 0.242 box mAP50-95 |
| DOTA8 OBB, after | 3/10 steps | 3.70e-4 | 5.557 | 0.332 box | 0.242 box mAP50-95 |

Additional loss movement matched the intended optimization change: COCO128 detection box loss fell 2.406 → 2.120; COCO128-seg segmentation loss fell 4.512 → 3.749. The 4-image DOTA8 validation set is too quantized to show a useful metric delta in ten epochs, despite the lower training losses. COCO128-seg mask mAP50-95 was effectively flat/slightly lower (0.142 → 0.141), so the evidence is not uniformly metric-positive; the consistent result is that training now reaches its intended LR and materially optimizes instead of remaining in warmup.

MPS emitted its existing nondeterministic `index_put_with_accumulate_mps` warning despite `deterministic=True`, so these are controlled local comparisons rather than bit-exact benchmark claims. Raw per-epoch CSVs are preserved locally in `/tmp/warmup-before-after/`.

**Comment by UltralyticsAssistant** [comment]
🎉 Thank you @glenn-jocher for this valuable contribution, now merged!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This warmup scheduling improvement makes training more predictable and resilient across short runs, dynamic batch-size adjustments, and time-based training. By ensuring regular training begins on schedule, it helps models converge more efficiently. We appreciate your continued efforts to make Ultralytics training workflows stronger and more reliable!
